### PR TITLE
[Darwin] Implemented download diagnostics log for MTRDevice_XPC

### PR DESCRIPTION
**Pull request by anush-apple**
- Implemented the downloadLogOfType API on MTRDevice_XPC to call into XPC server to forward the request
- Fixed the protocol to take controllerID which is missing from existing protocol

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDevice_XPC.mm [hunk 1]
```diff
@@ -269,6 +269,39 @@ - (void)_invokeCommandWithEndpointID:(NSNumber *)endpointID
...
+            MTR_LOG_ERROR("Error: %@", error);
+            dispatch_async(queue, ^{
+                completion(nil, [NSError errorWithDomain:MTRErrorDomain code:MTRErrorCodeGeneralError userInfo:nil]);
+            });
+        }] deviceController:[[self deviceController] uniqueIdentifier]
+                       nodeID:[self nodeID]
+            downloadLogOfType:type
+                      timeout:timeout
+                   completion:^(NSURL * _Nullable url, NSError * _Nullable error) {
+                       dispatch_async(queue, ^{
```
Comment (woody-apple) [inline src/darwin/Framework/CHIP/MTRDevice_XPC.mm:290]: If url is nil, we should likely turn this into an error...

File: src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h [hunk 2]
```diff
@@ -37,6 +37,10 @@ MTR_NEWLY_AVAILABLE
 - (oneway void)deviceController:(NSUUID *)controller nodeID:(NSNumber *)nodeID openCommissioningWindowWithSetupPasscode:(NSNumber *)setupPasscode discriminator:(NSNumber *)discriminator duration:(NSNumber *)duration completion:(MTRDeviceOpenCommissioningWindowHandler)completion;
 
 - (oneway void)downloadLogOfType:(MTRDiagnosticLogType)type nodeID:(NSNumber *)nodeID timeout:(NSTimeInterval)timeout completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion;
+
+@optional
+- (oneway void)deviceController:(NSUUID *)controller nodeID:(NSNumber *)nodeID downloadLogOfType:(MTRDiagnosticLogType)type timeout:(NSTimeInterval)timeout completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h:42]: Should probably document the ownership semantics of the file `url` points to here, when it's not nil.

File: src/darwin/Framework/CHIP/MTRDevice_XPC.mm (hunk 1 above)
Comment (anush-apple) [reply src/darwin/Framework/CHIP/MTRDevice_XPC.mm:290]: Looking at the implementation, download completion sets an NSError if the URL returned is nil. So the completion below is just sending that across.

File: src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h (hunk 2 above)
Comment (anush-apple) [reply src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h:45]: Added documentation

**Review by woody-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h [hunk 3]
```diff
@@ -37,6 +37,13 @@ MTR_NEWLY_AVAILABLE
 - (oneway void)deviceController:(NSUUID *)controller nodeID:(NSNumber *)nodeID openCommissioningWindowWithSetupPasscode:(NSNumber *)setupPasscode discriminator:(NSNumber *)discriminator duration:(NSNumber *)duration completion:(MTRDeviceOpenCommissioningWindowHandler)completion;
 
 - (oneway void)downloadLogOfType:(MTRDiagnosticLogType)type nodeID:(NSNumber *)nodeID timeout:(NSTimeInterval)timeout completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion;
+
+@optional
+/* Note: The consumer of the completion block should move the file that the url points to or open it for reading before the
+ * completion handler returns. Otherwise, the file will be deleted, and the data will be lost.
+ */
+- (oneway void)deviceController:(NSUUID *)controller nodeID:(NSNumber *)nodeID downloadLogOfType:(MTRDiagnosticLogType)type timeout:(NSTimeInterval)timeout completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/XPC Protocol/MTRXPCServerProtocol.h:45]: The comments here do not seem to match the code.  The code is doing an explicit remove, no?

**Pull request merged**